app/validators/status_reaction_validator.rb

```diff
@@ -0,0 +1,24 @@
+# frozen_string_literal: true
+
+class StatusReactionValidator < ActiveModel::Validator
+  SUPPORTED_EMOJIS = Oj.load_file(Rails.root.join('app', 'javascript', 'mastodon', 'features', 'emoji', 'emoji_map.json').to_s).keys.freeze
+
+  LIMIT = [1, (ENV['MAX_REACTIONS'] || 1).to_i].max
+
+  def validate(reaction)
+    return if reaction.name.blank?
+
+    reaction.errors.add(:name, I18n.t('reactions.errors.unrecognized_emoji')) if reaction.custom_emoji_id.blank? && !unicode_emoji?(reaction.name)
+    reaction.errors.add(:base, I18n.t('reactions.errors.limit_reached')) if reaction.account.local? && limit_reached?(reaction)
```

Comment:
```suggestion
    reaction.errors.add(:base, I18n.t('reactions.errors.limit_reached')) if reaction.account.local? && new_reaction?(reaction) && limit_reached?(reaction)
```
and adding:
```ruby
def new_reaction?(reaction)
    !reaction.status.status_reactions.where(status: reaction.status, account: reaction.account, name: reaction.name).exists?
end
```

This fixes the spec failure for the status_reaction fabricator I proposed in the other PR.

Reply:
The order doesn't really matter.
This just skips the validation if the reaction is not new, since you can't react with the same emoji twice anyway.

Reply:
The issue the fabricators_spec failure surfaced is, that the validator invalidates a reaction which already exists.
For this fix I oriented myself on the announcement reaction validator this validator is based on.